rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala

```diff
@@ -0,0 +1,850 @@
+        case NodePtr(nodePtrPrefix, ptr) =>
+          if (nodePtrPrefix.isEmpty) loadNode(ptr)
+          else
+            emptyNode.updated(byteToInt(nodePtrPrefix.head), NodePtr(nodePtrPrefix.tail, ptr)).pure
+      }
+
+    /**
+      * Optimize and save Node, create item from this Node
+      */
+    private def saveNodeAndCreateItem(
+        node: Node,
+        prefix: ByteVector,
+        compaction: Boolean = true
+    ): Item =
+      if (compaction) {
+        val nonEmptyItems = node.iterator.filter(_ != EmptyItem).take(2).toList
+        nonEmptyItems.size match {
+          case 0 => EmptyItem // All items are empty.
+          case 1 => // Only one item is not empty - merge child and parent nodes.
+            val idxItem = node.indexOf(nonEmptyItems.head)
+            nonEmptyItems.head match {
+              case EmptyItem => EmptyItem
+              case Leaf(leafPrefix, value) =>
+                Leaf(prefix ++ ByteVector(idxItem) ++ leafPrefix, value)
+              case NodePtr(nodePtrPrefix, ptr) =>
+                NodePtr(prefix ++ ByteVector(idxItem) ++ nodePtrPrefix, ptr)
+            }
+          case 2 => // 2 or more items are not empty.
+            NodePtr(prefix, saveNode(node))
+        }
+      } else NodePtr(prefix, saveNode(node))
+
+    /**
+      * Save new leaf value to this part of tree (start from curItems).
+      * Rehash and save all depend node to [[cacheW]].
+      *
+      * If exist leaf with same prefix but different value - update leaf value.
+      * If exist leaf with same prefix and same value - return [[None]].
+      * @return Updated current item.
+      */
+    def update(
+        curItem: Item,
+        insPrefix: ByteVector,
+        insValue: ByteVector
+    ): F[Option[Item]] =
+      curItem match {
+        case EmptyItem =>
+          (Leaf(insPrefix, insValue): Item).some.pure // Update EmptyItem to Leaf.
+
+        case Leaf(leafPrefix, leafValue) =>
+          assert(leafPrefix.size == insPrefix.size, "All Radix keys should be same length.")
+          if (leafPrefix == insPrefix) {
+            if (insValue == leafValue) none[Item].pure
+            else (Leaf(insPrefix, insValue): Item).some.pure
+          } // Update Leaf.
+          else {
+            // Create child node, insert existing and new leaf in this node.
+            // Intentionally not recursive for speed up.
+            val (commPrefix, insPrefixRest, leafPrefixRest) = commonPrefix(insPrefix, leafPrefix)
+            val newNode = emptyNode
+              .updated(byteToInt(leafPrefixRest.head), Leaf(leafPrefixRest.tail, leafValue))
+              .updated(byteToInt(insPrefixRest.head), Leaf(insPrefixRest.tail, insValue))
+            saveNodeAndCreateItem(newNode, commPrefix, compaction = false).some.pure
+          }
+
+        case NodePtr(ptrPrefix, ptr) =>
+          assert(ptrPrefix.size < insPrefix.size, "Radix key should be longer than NodePtr key.")
```

Comment:
For example, this assert can fail when function is called which is not expected by the caller because `F[_]` represents a _value_ which can be _executed_ multiple times (and possibly produce effects each time).